### PR TITLE
Prevent inputs ending up off screen or obscured by keyboards when linking from the error summary to inputs within a large fieldset

### LINKED ISSUE #1544: Error Summary link handling can result in obscured inputs
As reported by a user on Slack:

> I’ve got a bit of a weird one that only seems to affect iOS users. We have an address page that follows the standard pattern. When you click continue it will show the error messages which link to the appropriate inputs. If you click on the error message for the post code it seems to scroll down a bit but the input is covered my the iOS keyboard (please see attached screenshot).
>
> 
> image

The user confirmed that the address fields were inside a fieldset (as [we recommend in the addresses pattern](https://design-system.service.gov.uk/patterns/addresses/#multiple-text-inputs)).

I believe this is happening because the logic used when clicking on the error summary links is to scroll to the legend of the parent fieldset, if there is one. This makes sense for radios, checkboxes and the date input, but maybe less so in this situation.

### PR DESCRIPTION
**Pull request by 36degrees**
Only scroll from error summary to legend if the input is a radio or checkbox, or the input would end up in the top half of the screen.

This is to prevent linking to inputs which then end up either off the bottom of the screen or obscured by a software keyboard.

Fixes #1544

### CODE REVIEW

File: src/govuk/components/error-summary/error-summary.js [hunk 1]
```diff
@@ -109,7 +111,32 @@ ErrorSummary.prototype.getAssociatedLegendOrLabel = function ($input) {
     var legends = $fieldset.getElementsByTagName('legend')
 
     if (legends.length) {
-      return legends[0]
+      var $candidateLegend = legends[0]
```
Comment (nickcolley) [inline src/govuk/components/error-summary/error-summary.js]: In theory we could write tests for this logic, where we do using puppeteer on the demo pages you've been testing, this could be useful in the future if we (hopefully not) encounter other edge cases.

**Comment by nickcolley** [comment]
For reference there's also 'IntersectionObservers' which have concepts of thresholds, I bit overkill for this I think since we're only doing it once per link click but worth sharing: https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

File: src/govuk/components/error-summary/error-summary.js [hunk 2]
```diff
@@ -109,7 +111,32 @@ ErrorSummary.prototype.getAssociatedLegendOrLabel = function ($input) {
...
+      // screen, or obscured by a software keyboard.
+      var legendTop = $candidateLegend.getBoundingClientRect().top
+      var inputRect = $input.getBoundingClientRect()
+
+      // If the browser doesn't support Element.getBoundingClientRect().height
+      // or window.innerHeight (like IE8), bail and just link to the label.
+      if (inputRect.height && window.innerHeight) {
+        var inputBottom = inputRect.top + inputRect.height
+
+        if (inputBottom - legendTop < window.innerHeight / 2) {
```
Comment (nickcolley) [inline src/govuk/components/error-summary/error-summary.js]: When testing this on my machine I find that I'm always linked to the errored input for the last example despite the fieldset being able to fit all on the screen at once.

Is the issue that the keyboard opens after the navigation?

File: src/govuk/components/error-summary/error-summary.js (hunk 1 above)
Comment (36degrees) [reply src/govuk/components/error-summary/error-summary.js]: I've added to the existing tests to try and cover the two new examples that have been added to the review app.

Comment (nickcolley) [reply src/govuk/components/error-summary/error-summary.js]: Epic! Thank you

File: src/govuk/components/error-summary/error-summary.js (hunk 2 above)
Comment (36degrees) [reply src/govuk/components/error-summary/error-summary.js]: More or less – the logic checks to see whether the input would end up entirely in the top half of the viewport, to allow for keyboards.

**Comment by 36degrees** [comment]
@nickcolley happy with this changelog entry?

**Comment by nickcolley** [comment]
@36degrees 👍 looks peng mate
